File: 4-param-specs/basic_example.cpp
#include <concepts>
#include <iostream>
#include <string>
#include <utility>


template<typename C>
auto wrapper(C func) {
    auto inner = [func = std::move(func)]
            <typename... Args> requires std::invocable<C, Args...> (Args&&... args) -> decltype(auto) {
        return func(std::forward<Args>(args)...);
    };

    return inner;
}


float func(int i, std::string s) {
    std::cout << i << std::endl;
    std::cout << s << std::endl;
    return 3.14;
}


int main() {
    auto const wrapped_func = wrapper(func);
    auto const ret = wrapped_func(42, "foo");
}

File: 4-param-specs/basic_example.py
from typing import Callable, ParamSpec, TypeVar


T = TypeVar('T')
P = ParamSpec('P')


def wrapper(func: Callable[P, T]) -> Callable[P, T]:
    def inner(*args: P.args, **kwargs: P.kwargs) -> T:
        return func(*args, **kwargs)
    
    return inner


def func(i: int, s: str) -> float:
    print(i)
    print(s)
    return 3.14


wrapped_func = wrapper(func)
ret = wrapped_func(42, 'foo')
